### ISSUE TITLE
Temporary does not live long enough in let chains

### ISSUE THREAD

**Issue by Alexendoo** [rust-lang/rust#103476]
I tried this code, [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=af7f27fb4b68e713467f6f8a19802d35):

```rust
#![feature(let_chains)]

use syn::{Attribute, Meta};

fn f(attr: &Attribute) -> bool {
    if let Meta::NameValue(name_value) = attr.parse_meta().ok().unwrap()
        && let path_idents = name_value.path.segments.iter()
        /* other stuff */
    {
        
    }
    
    true
}
```

I expected to see this happen: code compiles

Instead, this happened: errors due to `name_value.path.segments does not live long enough`

It compiles when expressed as nested ifs, [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=96ab431660eb6144d52aeda9f645cc18):

```rust
#![feature(let_chains)]

use syn::{Attribute, Meta};

fn f(attr: &Attribute) -> bool {
    if let Meta::NameValue(name_value) = attr.parse_meta().ok().unwrap() {
        if let path_idents = name_value.path.segments.iter() {}
    }

    true
}
```

The code compiled previously, but no longer does after #103034 (confirmed with a `cargo bisect-rustc`)

cc @nathanwhit

**Comment by est31** [comment]
Minimized:

<details>

```Rust
#![feature(let_chains)]

struct Pd<T>(std::marker::PhantomData<T>);

impl<T> Pd<T> {
    fn iter(&self) -> Iter<T> {
        todo!()
    }
}

pub struct Iter<'a, T: 'a> {
    inner: Box<dyn IterTrait<'a, T> + 'a>,
}

trait IterTrait<'a, T: 'a> {
    fn clone_box(&self) -> Box<dyn IterTrait<'a, T> + 'a>;
}

fn f(m: Option<Pd<()>>) -> bool {
    if let Some(n) = m
        && let it = n.iter()
        /* other stuff */
    {
        
    }
    
    true
}
```

</details>

A bit further minimized (no syn dependency):

<details>

```Rust
#![feature(let_chains)]

struct Pd<T>(std::marker::PhantomData<T>);

impl<T> Pd<T> {
    fn iter(&self) -> Iter<T> {
        todo!()
    }
}

pub struct Iter<'a, T: 'a> {
    inner: Box<dyn IterTrait<'a, T> + 'a>,
}

trait IterTrait<'a, T: 'a> {
    fn clone_box(&self) -> Box<dyn IterTrait<'a, T> + 'a>;
}

fn f(m: Option<Pd<()>>) -> bool {
    if let Some(n) = m
        && let it = n.iter()
        /* other stuff */
    {
        
    }
    
    true
}
```

</details>

Last minimization step:

```Rust
#![feature(let_chains)]

struct Pd;

impl Pd {
    fn it(&self) -> It {
        todo!()
    }
}

pub struct It<'a>(Box<dyn Tr<'a>>);

trait Tr<'a> {}

fn f(m: Option<Pd>) {
    if let Some(n) = m && let it = n.iter() {};
}
```

**Comment by WaffleLapkin** [comment]
Even more minimized:

```rust
#![feature(let_chains)]
#![allow(irrefutable_let_patterns)]

fn it(_: &()) -> It<'_> { todo!() }

pub struct It<'a>(Box<dyn Tr<'a>>);

trait Tr<'a> {}

fn f() {
    if let n = () && let _ = it(&n) {};
}
```
```text
error[[E0597]](https://doc.rust-lang.org/nightly/error-index.html#E0597): `n` does not live long enough
  --> src/lib.rs:11:33
   |
11 |     if let n = () && let _ = it(&n) {};
   |                              ---^^-  -- ... and the borrow might be used here, when that temporary is dropped and runs the destructor for type `It<'_>`
   |                              |  |    |
   |                              |  |    `n` dropped here while still borrowed
   |                              |  borrowed value does not live long enough
   |                              a temporary with access to the borrow is created here ...
```
[[play]](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=2022b804600432f7f52f40f464bb77ff)

----

Weirdly, I couldn't replace `Box<dyn Tr<'a>>` with anything else, I thought that it's just significant `Drop` & invariant lifetime `'a`, but something like `String, fn(&'a ()) -> &'a ()` compiles, so I'm not sure what's going on.

**Comment by alercah** [comment]
Box's `drop` is actually hard-coded magic, so it might be unique to it.

**Comment by est31** [comment]
Good point about the lang item. I've tried it with `Arc` and `Vec`, it seems to fail for those too.

This also still reproduces:

```Rust
#![feature(let_chains)]
#![allow(irrefutable_let_patterns)]

struct B<T: ?Sized> { _t: std::marker::PhantomData<T>, }

impl<T: ?Sized> std::ops::Drop for B<T> {
    fn drop(&mut self) {}
}

fn it(_: &()) -> B<dyn Tr<'_>> { todo!() }

trait Tr<'a> {}

fn f() {
    if let n = () && let _ = it(&n) {};
}
```

**Comment by nathanwhit** [comment]
This reproduces as well:
```rust
#![feature(let_chains)]
#![allow(irrefutable_let_patterns)]

struct B<'a>(&'a ());

impl<'a> std::ops::Drop for B<'a> {
    fn drop(&mut self) {}
}

fn f() {
    if let n = () && let _ = B(&n) {};
}
```

If you remove the `Drop` impl or the lifetime parameter it compiles without error.

**Comment by WaffleLapkin** [comment]
The reason why `Box<&'a ()>` / `Vec<&'a ()>` don't work is probably because of `#[may_dangle]`.

**Comment by est31** [comment]
There is no `#[may_dangle]` in the two last snippets, so I think it's not because of those.

**Comment by WaffleLapkin** [comment]
@est31 that's the point? @nathanwhit's example can use `&'a ()` because the drop doesn't have `#[may_dangle]`. `Arc` has `#[may_dangle]`, so `Arc<&'a ()>` does not reproduce the issue (you need a `dyn Trait` because it's possibly drop and has the lifetime).

**Comment by est31** [comment]
Good point, you are right about `#[may_dangle]` I think. If you remove the unsafe from and `#[may_dangle]` from here it fails again (but works right now):

```Rust
#![feature(dropck_eyepatch)]
#![feature(let_chains)]
#![allow(irrefutable_let_patterns)]

struct B<T: ?Sized> { _t: *const T, }

unsafe impl<#[may_dangle] T: ?Sized> std::ops::Drop for B<T> {
    fn drop(&mut self) {}
}

fn it(_: &()) -> B<&()> { todo!() }

trait Tr<'a> {}

fn f() {
    if let n = () && let _ = it(&n) {};
}
```

**Comment by WiSaGaN** [comment]
Does this still block the stabilization of let chains?

**Comment by est31** [comment]
This doesn't reproduce on edition 2024 any more, while still reproducing on edition 2021. I would say it's fixed by if let rescoping (#124085).

I've tried both the ["a bit further minimized"](https://github.com/rust-lang/rust/issues/103476#issuecomment-1296338364) snippet as well as the [snippet right below](https://github.com/rust-lang/rust/issues/103476#issuecomment-1298155838) by @WaffleLapkin.

As to the original snippet provided by @Alexendoo , same behaviour there, although note that latest syn (v1.0.109) does not reproduce the issue while an older version like 1.0.80 reproduces it.

I suppose this can be closed now once we have a test for it in the testsuite.
